man/tpm2_startauthsession.1.md
% tpm2_startauthsession(1) tpm2-tools | General Commands Manual

# NAME

**tpm2_startauthsession**(1) - Start a session with the TPM.

# SYNOPSIS

**tpm2_startauthsession** [*OPTIONS*]

# DESCRIPTION

**tpm2_startauthsession**(1) - Starts a session with the TPM. The default is to
start a *trial* session unless the **-a** option is specified. Saves the policy
session data to a file. This file can then be used in subsequent tools that can
use a policy file for authorization or policy events.

This will not work with resource managers (RMs) outside of
[tpm2-abrmd](https://github.com/tpm2-software/tpm2-abrmd), as most RMs will flush
session handles when a client disconnects from the IPC channel. However, when using
a RM without the session gapping feature, one can use the command TCTI to keep the
connection open.

The first step is to create a socket listener that uses tpm2\_send:
```bash
mknod "$HOME/backpipe" p
while [ 1 ]; do tpm2_send 0<"$HOME/backpipe" | nc -lU "$HOME/sock" 1>"$HOME/backpipe"; done;
```

The next step is to use the command TCTI and netcat (nc) to send data to the socket.
```bash
tpm2_startauthsession --tcti="cmd:nc -q 0 -U $HOME/sock" <options>
```

When finishing ensure to kill the listener. For commands executed with the command tcti against
the listener, one will need to manage transient handles. The simplest way is to add a flush
after each command: `tpm2_flushcontext --tcti="cmd:nc -q 0 -U $HOME/sock" -t`

Note: This example uses UNIX sockets, since the socket is controlled with Linux
access controls. Using a port is not recommended as it's either open to any user
on the system (localhost) or bound to a network card and exposed to the network.

This will work with direct TPM access, but note that internally this calls a
*ContextSave* and a *ContextLoad* on the session handle, thus the session
**cannot** be saved/loaded again.

# OPTIONS

  * **\--policy-session**:

    Start a policy session of type **TPM_SE_POLICY**. Default without this
    option is **TPM_SE_TRIAL**.

    **NOTE**: A *trial* session is used when building a policy and a *policy*
    session is used when authenticating with a policy.

  * **\--audit-session**:

    Start an HMAC session to be used as an audit session. Default without
    this option is **TPM2_SE_TRIAL**.

  * **\--hmac-session**:

    Start an HMAC session of type **TPM_SE_HMAC**. Default without this option
    is **TPM2_SE_TRIAL**.

  * **-g**, **\--hash-algorithm**=_ALGORITHM_:

    The hash algorithm used in computation of the policy digest.

  * **-G**, **\--key-algorithm**=_ALGORITHM_:

    The symmetric algorithm used in parameter encryption/decryption.

  * **-c**, **\--key-context**=_OBJECT_:

    Set the tpmkey and bind objects to be the same.
    Session parameter encryption is turned on.
    Session parameter decryption is turned on.
    Parameter encryption/decryption symmetric-key set to AES-CFB.

  * **-S**, **\--session**=_FILE_:

    The name of the policy session file, required.

  * **\--bind-context**=_FILE_:

    Set the bind object.
    Session parameter encryption is off. Use **tpm2_sessionconfig** to turn on.
    Session parameter decryption is off. Use **tpm2_sessionconfig** to turn on.
    Parameter encryption/decryption symmetric-key set to AES-CFB.

  * **\--bind-auth**=_AUTH_:

    Set the authorization value for the bind object.

  * **\--tpmkey-context**=_FILE_:

    Set the tpmkey object.
    Session parameter encryption is off. Use **tpm2_sessionconfig** to turn on.
    Session parameter decryption is off. Use **tpm2_sessionconfig** to turn on.
    Parameter encryption/decryption symmetric-key set to AES-CFB.

  * **-n**, **\--name**=_FILE_

    A name file as output from a tool like tpm2\_readpublic(1) `-n` option.
    The name file can be used to **verify** a persistent handle input for
    the `--tpmkey-context`, `-c`, and `--key-context` options. Verification
    that the object referenced by a peristent handle, e.g 0x81000000, is
    the key expected prevents attackers from performing a man-in-the-middle
    attack on session traffic.

## References

[context object format](common/ctxobj.md) details the methods for specifying
_OBJECT_.

[authorization formatting](common/authorizations.md) details the methods for
specifying _AUTH_.

[common options](common/options.md) collection of common options that provide
information many users may expect.

[common tcti options](common/tcti.md) collection of options used to configure
the various known TCTI modules.

# EXAMPLES

## Parameter encryption.

Command parameters and responses are not encrypted if no hmac session is used for
commands. A session where the parameter, in this case the password, is
encrypted can be created as follows:

```bash
tpm2_createprimary -c prim.ctx
tpm2_startauthsession -S enc_session.ctx --hmac-session --key-context prim.ctx
tpm2_create -C prim.ctx -c key.ctx -S enc_session.ctx -p password
```
Even without using a key for the session, a session can be used for parameter
encryption:

```bash
tpm2_startauthsession -S enc_session.ctx --hmac-session
tpm2_sessionconfig enc_session.ctx --enable-decrypt
tpm2_create -C prim.ctx -c key.ctx -S enc_session.ctx -p password
```
If the response has to be encrypted, the --enable-encrypt parameter can be used:

```bash
tpm2_startauthsession -S enc_session.ctx --hmac-session
tpm2_sessionconfig enc_session.ctx --enable-encrypt
tpm2_getrandom -S enc_session.ctx -o rand
```

## Start a *trial* session and save the session data to a file
```bash
tpm2_startauthsession -S mysession.ctx
```

## Start a *policy* session and save the session data to a file
```bash
tpm2_startauthsession --policy-session -S mysession.ctx
```

## Start an encrypted and bound *policy* session and save the session data to a file
```bash
tpm2_createprimary -c primary.ctx
tpm2_startauthsession --policy-session -c primary.ctx -S mysession.ctx
```

[returns](common/returns.md)

[footer](common/footer.md)


CHANGELOG
v2.2.2 - next:
*  fix make distcheck
v2.1.1 - next:
* Potential memory leak fix when tcti/sapi initialization fails.
* tpm2_listpcrs: use TPM2_GetCapability to determine PCRs to read
* listpcrs: remove one redundant call to tpm get cap
* listpcrs: fix for unsupported/disabled alg in -L
* build: use supported comment to suppress GCC7 fallthrough warning
* kdfa: allow to build with OpenSSL 1.1.x

v2.1.0 - 2017-7-24
* Fix readx and writex on multiple EINTR returns.
* Add support for the tabrmd TCTI. This is the new default.
* Change default socket port from 2323 (the old resourcemgr) to 2321 (default
  simulator port).
* Cherry-pick fix for CVE-2017-7524.
* Fix tpm2_listpcr command line option handling.
* Fix tpm2_getmanufec memory issues.

v2.0 - 2017-03-29
* Tracked on the milestone: https://github.com/01org/tpm2.0-tools/milestone/2
* Reworked all the tools to support configurable TCTIs, based on build time
  configuration, one can specify the tcti via the --tcti (-T) option to all
  tools.
* tpm2_getrandom interface made -s a positional argument.
* Numerous bug fixes.

v1.1 - 2016-11-04
Added
* travis ci support.
* Allow for unit tests to be enabled selectively.
* tpm2_rc_decode tool: Decode TPM_RC error codes.
* Android Make file
* tpm2_listpersistent: list all persistent objects
* test scripts for tpm2-tools
* tpm2_nvreadlock
* tpm2_getmanufec: retrieve EC from tpm manufacturer server.
* Copy 'common' and 'sample' code from the TPM2.0-TSS repo.

Modified
* tpm2_takeownership: update option -c to use lockout password to clear.
* tpm2_listpcrs: add options -L and -s, rewrite to increase performance.
* tpm2_quote: added -L option to support selection of multiple banks.
* tpm2_quote: add -q option to get qualifying data.
* configure: Use pkg-config to get info about libcurl and libcrypto.
* configure: Use pkg-config to locate SAPI and TCTI headers / libraries.
* tpm2_x: Add -X option to enable password input in Hex format.
* tpm2_nvdefine: Change -X option to -I.
* tpm2-nvwrite: fix for unable to write 1024B+ data.
* tpm2_getmanufec: Fix base64 encoding.
* tpm2_x: fixed a lot of TPM2B failures caused by wrong initialization.
* tpm2_getmanufec: let configure handle libs.
* tpm2_getmanufec: Convert from dos to unix format.
* build: Check for TSS2 library @ configure time.
* build: Detect required TSS2 and TCTI headers.
* build: Use libtool to build the common library
* build: Install all binaries into sbin.
* build: Build common sources into library.
* build: Move all source files to 'src'.
* Makefile.am: Move all build rules into single Makefile.am.
* everything: Use new TCTI headers and fixup API calls.
* everything: Update source to cope with sapi header cleanup.
* tpm2_activatecredential: Updated to support TCG compatible EK
* tpm2_getpubak: Updated to use TCG compatible EK
* tpm2_getpubek: fix ek creation to follow TCG EK profile spec.

Removed
* Windows related code
* depenedency on the TPM2.0-TSS repo source code

v1.0 - 2015-10-19
* 1.0 release
* 29 tools included


tools/tpm2_startauthsession.c
/* SPDX-License-Identifier: BSD-3-Clause */

#include <stdbool.h>
#include <string.h>

#include "files.h"
#include "log.h"
#include "object.h"
#include "tpm2.h"
#include "tpm2_auth_util.h"
#include "tpm2_tool.h"
#include "tpm2_alg_util.h"
#include "tpm2_options.h"

typedef struct tpm2_startauthsession_ctx tpm2_startauthsession_ctx;
struct tpm2_startauthsession_ctx {
    struct {
        TPM2_SE type;
        TPMI_ALG_HASH halg;
        TPMT_SYM_DEF sym;
        struct {
            /*
             * Salt generated by esys is encrypted using the tpmkey and
             * encryption does not require auth to be specified.
             */
            const char *key_context_arg_str;
            tpm2_loaded_object key_context_object;
        } tpmkey;

        struct {
            /*
             * While TPM2_CC_StartAuthSession does not required the auth of the
             * bind to be specified, it is captured here for esys to calculate
             * the sessionkey.
             */
            const char *bind_context_arg_str;
            const char *bind_context_auth_str;
            tpm2_loaded_object bind_context_object;
        } bind;
    } session;
    struct {
        const char *path;
    } output;

    tpm2_session_data *session_data;
    TPMA_SESSION attrs;
    bool is_real_policy_session;
    bool is_hmac_session;
    bool is_session_encryption_possibly_needed;
    bool is_session_audit_required;
    /* Salted/ Bounded session combinations */
    bool is_salted;
    bool is_bounded;
    bool is_salt_and_bind_obj_same;

    const char *name_path;
    TPM2B_NAME name;
};

static tpm2_startauthsession_ctx ctx = {
    .attrs = TPMA_SESSION_CONTINUESESSION,
    .session = {
        .type = TPM2_SE_TRIAL,
        .halg = TPM2_ALG_SHA256,
        .sym = {
            .algorithm = TPM2_ALG_AES,
            .keyBits = { .aes = 128 },
            .mode = { .aes = TPM2_ALG_CFB }
        },
    },
    .name = {
        .size = BUFFER_SIZE(TPM2B_NAME, name)
    },
};

static bool on_option(char key, char *value) {

    switch (key) {
    case 0:
        ctx.is_real_policy_session = true;
        break;
    case 1:
        ctx.is_hmac_session = true;
        ctx.is_session_audit_required = true;
        ctx.attrs |= TPMA_SESSION_AUDIT;
        break;
    case 'g':
        ctx.session.halg = tpm2_alg_util_from_optarg(value,
                tpm2_alg_util_flags_hash);
        if (ctx.session.halg == TPM2_ALG_ERROR) {
            LOG_ERR("Invalid choice for policy digest hash algorithm");
            return false;
        }
        break;
    case 'G':
        ctx.session.sym.algorithm = tpm2_alg_util_from_optarg(value,
                tpm2_alg_util_flags_symmetric);
        if (ctx.session.sym.algorithm == TPM2_ALG_ERROR) {
            LOG_ERR("Invalid symmetric algorithm");
            return false;
        }
        break;
    case 'S':
        ctx.output.path = value;
        break;
    case 'c':
        ctx.is_salt_and_bind_obj_same = true;
        ctx.session.tpmkey.key_context_arg_str = value;
        ctx.session.bind.bind_context_arg_str = value;
        ctx.is_session_encryption_possibly_needed = true;
        ctx.attrs |= (TPMA_SESSION_DECRYPT | TPMA_SESSION_ENCRYPT);
        break;
    case 'n':
        ctx.name_path = value;
        break;
    case 2:
        ctx.is_bounded = true;
        ctx.session.bind.bind_context_arg_str = value;
        ctx.is_session_encryption_possibly_needed = true;
        break;
    case 3:
        ctx.session.bind.bind_context_auth_str = value;
        break;
    case 4:
        ctx.is_salted = true;
        ctx.session.tpmkey.key_context_arg_str = value;
        ctx.is_session_encryption_possibly_needed = true;
        break;
    case 5:
        ctx.is_hmac_session = true;
        ctx.is_session_encryption_possibly_needed = true;
        break;
    }

    return true;
}

static bool tpm2_tool_onstart(tpm2_options **opts) {

    static struct option topts[] = {
        { "policy-session", no_argument,       NULL,  0 },
        { "audit-session",  no_argument,       NULL,  1 },
        { "bind-context",   required_argument, NULL,  2 },
        { "bind-auth",      required_argument, NULL,  3 },
        { "tpmkey-context", required_argument, NULL,  4 },
        { "hmac-session",   no_argument,       NULL,  5 },
        { "hash-algorithm", required_argument, NULL, 'g'},
        { "key-algorithm",  required_argument, NULL, 'G'},
        { "session",        required_argument, NULL, 'S'},
        { "key-context",    required_argument, NULL, 'c'},
        { "name",           required_argument, NULL, 'n'},
    };

    *opts = tpm2_options_new("g:G:S:c:n:", ARRAY_LEN(topts), topts, on_option,
    NULL, 0);

    return *opts != NULL;
}

static tool_rc is_input_options_valid(void) {

    if (!ctx.output.path) {
        LOG_ERR("Expected option -S");
        return tool_rc_option_error;
    }

    /* Trial-session: neither real_policy nor audit/hmac session */
    if (!ctx.is_real_policy_session && !ctx.is_hmac_session &&
    ctx.is_session_encryption_possibly_needed) {
        LOG_ERR("Trial sessions cannot be additionally used as encrypt/decrypt "
                "session");
        return tool_rc_option_error;
    }

    /*
     * Only use --key-context if both bind and tpmkey objects are the same.
     */
    if (ctx.is_salted && ctx.is_salt_and_bind_obj_same) {
        LOG_ERR("Specify --key-context or tpmkey-context, not both.");
        return tool_rc_option_error;
    }

    if (ctx.is_bounded && ctx.is_salt_and_bind_obj_same) {
        LOG_ERR("Specify --key-context or --bind-context, not both.");
        return tool_rc_option_error;
    }

    if (ctx.session.bind.bind_context_auth_str &&
    !ctx.session.bind.bind_context_arg_str) {
        LOG_ERR("Specify the bind entity when specifying the bind auth "
                "even when bind is same as tpmkey object.");
        return tool_rc_option_error;
    }

    /*
     * Setting sessions for audit should be handled with tpm2_sessionconfig
     * The following support is for backwards compatibility
     */
    if (ctx.is_real_policy_session && ctx.is_session_audit_required) {
        LOG_ERR("Policy sessions cannot be additionally used for audit");
        return tool_rc_option_error;
    }

    /*
     * A session cannot be without a purpose.
     */
    if (!(ctx.attrs & TPMA_SESSION_AUDIT) &&
    !(ctx.attrs & TPMA_SESSION_ENCRYPT) &&
    !(ctx.attrs & TPMA_SESSION_DECRYPT) && ctx.is_hmac_session) {
        LOG_WARN("Session has to be used either for auth and/or audit and/or "
                 "parameter-encryption/decryption. Use session-config tool to "
                 "specify the use");
    }

    return tool_rc_success;
}

static tool_rc setup_session_data(void) {

    if (ctx.is_real_policy_session) {
        ctx.session.type = TPM2_SE_POLICY;
    }

    if (ctx.is_hmac_session) {
        ctx.session.type = TPM2_SE_HMAC;
    }

    ctx.session_data = tpm2_session_data_new(ctx.session.type);
    if (!ctx.session_data) {
        LOG_ERR("oom");
        return tool_rc_general_error;
    }

    tpm2_session_set_path(ctx.session_data, ctx.output.path);

    tpm2_session_set_authhash(ctx.session_data, ctx.session.halg);

    if (ctx.is_session_encryption_possibly_needed) {

        tpm2_session_set_symmetric(ctx.session_data, &ctx.session.sym);
    }

    if (ctx.session.bind.bind_context_arg_str) {
        tpm2_session_set_bind(ctx.session_data,
        ctx.session.bind.bind_context_object.tr_handle);
    }

    if (ctx.session.tpmkey.key_context_arg_str) {
        tpm2_session_set_key(ctx.session_data,
        ctx.session.tpmkey.key_context_object.tr_handle);
    }

    tpm2_session_set_attrs(ctx.session_data, ctx.attrs);

    return tool_rc_success;
}

static tool_rc process_input_data(ESYS_CONTEXT *ectx) {

    if (ctx.name_path) {
        bool ret = files_load_bytes_from_path(ctx.name_path, ctx.name.name, &ctx.name.size);
        if (!ret) {
            LOG_ERR("Could load name from path: \"%s\"", ctx.name_path);
            return tool_rc_general_error;
        }
    }

    /*
     * Backwards compatibility behavior/ side-effect:
     *
     * The presence of a tpmkey and bind object should not result in setting up
     * the session for parameter encryption. It is not a requirement. IOW one
     * can have a salted and bounded session and not perform parameter
     * encryption.
     */

    if (ctx.session.tpmkey.key_context_arg_str) {
    /*
     * attempt to set up the encryption parameters for this, we load an ESYS_TR
     * from disk for transient objects and we load from tpm public for
     * persistent objects. Deserialized ESYS TR objects are checked.
     */
        tool_rc rc = tpm2_util_object_load(ectx,
                ctx.session.tpmkey.key_context_arg_str,
                &ctx.session.tpmkey.key_context_object, TPM2_HANDLE_ALL_W_NV);
        if (rc != tool_rc_success) {
            return rc;
        }

        /* if loaded object is non-permanant, it should ideally be persistent */
        if (ctx.session.tpmkey.key_context_object.handle) {

            bool is_transient = (ctx.session.tpmkey.key_context_object.handle
                    >> TPM2_HR_SHIFT) == TPM2_HT_TRANSIENT;
            if (!is_transient && !ctx.name_path) {
                LOG_WARN("check public portion of the tpmkey manually");
            }

            /*
             * ESAPI performs this check when an ESYS_TR or Context file is used, so we
             * could only run the check on the case where a raw TPM handle is provided,
             * however, it seems prudent that if the user specifies a name, we always
             * just check it.
             */
            if (ctx.name_path) {
                TPM2B_NAME *got_name = NULL;
                rc = tpm2_tr_get_name(ectx, ctx.session.tpmkey.key_context_object.tr_handle,
                        &got_name);
                if (rc != tool_rc_success) {
                    return rc;
                }

                bool is_expected = cmp_tpm2b(name, &ctx.name, got_name);
                Esys_Free(got_name);
                if (!is_expected) {
                    LOG_ERR("Expected name does not match");
                    return tool_rc_general_error;
                }
            }
        }
    }

    /*
     * We need to load the bind object and set its auth value in the bind
     * objects ESYS_TR.
     *
     * A loaded object creates another session and that is not what we want.
     */
    if (ctx.session.bind.bind_context_arg_str) {
        tool_rc rc = tpm2_util_object_load(ectx,
                ctx.session.bind.bind_context_arg_str,
                &ctx.session.bind.bind_context_object, TPM2_HANDLE_ALL_W_NV);
        if (rc != tool_rc_success) {
            return rc;
         }
    }

    if (ctx.session.bind.bind_context_auth_str) {
        TPM2B_AUTH authvalue = { 0 };
        bool result = handle_password(
            ctx.session.bind.bind_context_auth_str, &authvalue);
        if (!result) {
            return tool_rc_general_error;
        }

        tool_rc rc = tpm2_tr_set_auth(ectx,
        ctx.session.bind.bind_context_object.tr_handle, &authvalue);
        if (rc != tool_rc_success) {
            LOG_ERR("Failed setting auth in the bind object ESYS_TR");
            return rc;
        }
    }

    return setup_session_data();
}

static tool_rc tpm2_tool_onrun(ESYS_CONTEXT *ectx, tpm2_yaml *doc, tpm2_option_flags flags) {

    UNUSED(flags);

    //Check input options
    tool_rc rc = is_input_options_valid();
    if (rc != tool_rc_success) {
        return rc;
    }

    //Process inputs
    rc = process_input_data(ectx);
    if (rc != tool_rc_success) {
        return rc;
    }

    //ESAPI call to start session
    tpm2_session *s = NULL;
    rc = tpm2_session_open(ectx, ctx.session_data, &s);
    if (rc != tool_rc_success) {
        return rc;
    }

    //Process outputs
    return tpm2_session_close(&s);
}

// Register this tool with tpm2_tool.c
TPM2_TOOL_REGISTER("startauthsession", tpm2_tool_onstart, tpm2_tool_onrun, NULL,
NULL)
